### PR TITLE
add configuration to limit replication workers

### PR DESCRIPTION
**Pull request by poornas**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Motivation and Context
Provide option to cap number of replication workers

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: internal/config/api/api.go [hunk 1]
```diff
@@ -31,13 +31,15 @@ import (
...
-	apiListQuorum                  = "list_quorum"
-	apiReplicationPriority         = "replication_priority"
+	apiRequestsMax             = "requests_max"
+	apiRequestsDeadline        = "requests_deadline"
+	apiClusterDeadline         = "cluster_deadline"
+	apiCorsAllowOrigin         = "cors_allow_origin"
+	apiRemoteTransportDeadline = "remote_transport_deadline"
+	apiListQuorum              = "list_quorum"
+	apiReplicationPriority     = "replication_priority"
+	apiReplicationMaxWorkers   = "replication_workers_max"
```
Comment (harshavardhana) [inline internal/config/api/api.go]:
```suggestion
	apiReplicationMaxWorkers   = "replication_max_workers"
```

File: internal/config/api/api.go [hunk 2]
```diff
@@ -48,16 +50,16 @@ const (
...
+	EnvAPIRequestsMax                 = "MINIO_API_REQUESTS_MAX"
+	EnvAPIRequestsDeadline            = "MINIO_API_REQUESTS_DEADLINE"
+	EnvAPIClusterDeadline             = "MINIO_API_CLUSTER_DEADLINE"
+	EnvAPICorsAllowOrigin             = "MINIO_API_CORS_ALLOW_ORIGIN"
+	EnvAPIRemoteTransportDeadline     = "MINIO_API_REMOTE_TRANSPORT_DEADLINE"
+	EnvAPITransitionWorkers           = "MINIO_API_TRANSITION_WORKERS"
+	EnvAPIListQuorum                  = "MINIO_API_LIST_QUORUM"
+	EnvAPISecureCiphers               = "MINIO_API_SECURE_CIPHERS" // default config.EnableOn
+	EnvAPIReplicationPriority         = "MINIO_API_REPLICATION_PRIORITY"
+	EnvAPIReplicationWorkersMax       = "MINIO_API_REPLICATION_MAX_WORKERS"
```
Comment (harshavardhana) [inline internal/config/api/api.go]:
```suggestion
	EnvAPIReplicationMaxWorkers       = "MINIO_API_REPLICATION_MAX_WORKERS"
```

File: internal/config/api/api.go [hunk 3]
```diff
@@ -156,6 +162,7 @@ type Config struct {
 	RemoteTransportDeadline     time.Duration `json:"remote_transport_deadline"`
 	ListQuorum                  string        `json:"list_quorum"`
 	ReplicationPriority         string        `json:"replication_priority"`
+	ReplicationMaxWorkers       int           `json:"replication_workers_limit"`
```
Comment (harshavardhana) [inline internal/config/api/api.go]:
```suggestion
	ReplicationMaxWorkers       int           `json:"replication_max_workers"`
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

File: cmd/handler-api.go [hunk 4]
```diff
@@ -334,15 +336,21 @@ func maxClients(f http.HandlerFunc) http.HandlerFunc {
...
-func (t *apiConfig) getReplicationPriority() string {
+func (t *apiConfig) getReplicationOpts() replicationPoolOpts {
 	t.mu.RLock()
 	defer t.mu.RUnlock()
 
 	if t.replicationPriority == "" {
-		return "auto"
+		return replicationPoolOpts{
+			Priority:   "auto",
+			MaxWorkers: 500,
```
Comment (shtripat) [inline cmd/handler-api.go]: Should we define a const for this max value of 500?

File: cmd/bucket-replication.go [hunk 5]
```diff
@@ -2024,21 +2044,24 @@ func (p *ReplicationPool) queueReplicaTask(ri ReplicateObjectInfo) {
...
+				workers := int(math.Min(float64(len(p.workers)+1), float64(maxWorkers)))
 				existing := len(p.workers)
 				p.mu.RUnlock()
 				p.ResizeWorkers(workers, existing)
 			}
-			if p.ActiveMRFWorkers() < MRFWorkerMaxLimit {
+			maxMRFWorkers := int(math.Min(float64(maxWorkers), MRFWorkerMaxLimit))
+			if p.ActiveMRFWorkers() < maxMRFWorkers {
 				p.mu.RLock()
 				workers := int(math.Min(float64(p.mrfWorkerSize+1), MRFWorkerMaxLimit))
```
Comment (krisis) [inline cmd/bucket-replication.go]: Shouldn't this be,
```suggestion
				workers := int(math.Min(float64(p.mrfWorkerSize+1), maxMRFWorkers))
```

**Review by krisis — APPROVED** [review]

**Pull request merged**